Add Terms and Conditions (T&C) Feature Support for Commissioning

swan-amazon opened the pull request:
This commit introduces comprehensive support for handling Terms and Conditions (T&C) during the commissioning process, enhancing compatibility with Matter device certification requirements.

Key changes include:

1. **Commissioning Process Updates**:
   - Introduced `SetRequireTermsAndConditionsAcknowledgement`, `SetTermsAcknowledgements`, and `SetSkipCommissioningComplete` APIs in the commissioning library.
   - Updated commissioning stages to include `kGetTCAcknowledgments` and `kConfigureTCAcknowledgments` for seamless integration of T&C acknowledgements.
   - Added methods for processing T&C acknowledgements and advancing commissioning stages upon user response.

2. **Test Framework Enhancements**:
   - Added arguments (`tc_version`, `tc_user_response`, `in_test_commissioning_method`) for specifying T&C configuration in tests.
   - Enhanced `populate_commissioning_args` to manage new T&C-related arguments.
   - Updated Python test bindings and Matter test infrastructure to support T&C workflows.

3. **Chip-Tool Improvements**:
   - Extended `PairingCommand` to handle T&C-related arguments (`require-tc-acknowledgements`, `tc-acknowledgements`, `tc-acknowledgements-version`) for test scenarios.
   - Ensured backward compatibility by defaulting new parameters to preserve pre-1.4 behavior.

swan-amazon committed:
Add Terms and Conditions (T&C) Feature Support for Commissioning

This commit introduces comprehensive support for handling Terms and
Conditions (T&C) during the commissioning process, enhancing
compatibility with Matter device certification requirements.

Key changes include:

1. **Commissioning Process Updates**:
   - Introduced `SetRequireTermsAndConditionsAcknowledgement`,
     `SetTermsAcknowledgements`, and `SetSkipCommissioningComplete` APIs
     in the commissioning library.
   - Updated commissioning stages to include `kGetTCAcknowledgments` and
     `kConfigureTCAcknowledgments` for seamless integration of T&C
     acknowledgements.
   - Added methods for processing T&C acknowledgements and advancing
     commissioning stages upon user response.

2. **Test Framework Enhancements**:
   - Added arguments (`tc_version`, `tc_user_response`,
     `in_test_commissioning_method`) for specifying T&C configuration in
     tests.
   - Enhanced `populate_commissioning_args` to manage new T&C-related
     arguments.
   - Updated Python test bindings and Matter test infrastructure to
     support T&C workflows.

3. **Chip-Tool Improvements**:
   - Extended `PairingCommand` to handle T&C-related arguments
     (`require-tc-acknowledgements`, `tc-acknowledgements`,
     `tc-acknowledgements-version`) for test scenarios.
   - Ensured backward compatibility by defaulting new parameters to
     preserve pre-1.4 behavior.

bzbarsky-apple commented on examples/chip-tool/commands/pairing/PairingCommand.h:
```diff
@@ -202,6 +202,18 @@ class PairingCommand : public CHIPCommand,
                         Argument::kOptional);
+
+            AddArgument("require-tc-acknowledgements", 0, 1, &mRequireTCAcknowledgements,
+                        "Terms and Conditions acknowledgements is known to be required or not, (when required, the commissioner "
+                        "will wait until they are provided)");
```
This does not seem like the right description, in that there is no waiting; when required you can provide them via the other args, right?

Also descriptions for optional args should say what the default behavior is.

swan-amazon replied:
I updated the description each of the three new arguments.

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -3207,6 +3233,52 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
+            return;
+        }
+
+        ChipLogProgress(Controller, "Waiting for Terms and Conditions");
```
Doesn't this need to extend the fail-safe or something?  What happens if the user takes 10 minutes to read the terms (which is a bare minimum for how long it can possibly take to actually read those things typically)?

swan-amazon replied:
We talked offline. I removed the wait-state. The user input related to T&C acknowledgements availability now needs to be a precondition to starting AutoCommissioner.

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -706,6 +706,25 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+    /**
+     * @brief
+     *   This function is called by the upper layer application to indicate that the required terms and conditions
+     *   acknowledgements have been set. This function should be called after the terms and conditions bitmask and version
```
"Have been set" where and how?

swan-amazon replied:
After discussing offline, I have removed the wait-state.

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -706,6 +706,25 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+     * @brief
+     *   This function is called by the upper layer application to indicate that the required terms and conditions
+     *   acknowledgements have been set. This function should be called after the terms and conditions bitmask and version
+     *   have been defined using the appropriate configuration macros and the application has gathered the necessary
```
Which configuration macros, exactly?

swan-amazon replied:
The function and comment block have been removed.

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -706,6 +706,25 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+     *   have been defined using the appropriate configuration macros and the application has gathered the necessary
+     *   acknowledgements from the user.
+     *
+     * The upper layer application should call this method once it has successfully presented and obtained acknowledgements
```
How does the upper layer application know it needs to do that?

swan-amazon replied:
After removing the wait-state, this function is no longer required. So, the comment is no longer applicable.

bzbarsky-apple commented on src/controller/CommissioningDelegate.h:
```diff
@@ -81,7 +80,10 @@ enum CommissioningStage : uint8_t
+    kRemoveThreadNetworkConfig,       ///< Remove Thread network config.
+    kGetTCAcknowledgments,            ///< Waiting for the higher layer to provide terms and conditions acknowledgements.
+    kConfigureTCAcknowledgments,      ///< Send SetTCAcknowledgements (0x30:6) command to the device
+    kCleanup,                         ///< Call delegates with status, free memory, clear timers and state
```
The comments above claim this value cannot be moved later in the list.  Either it _can_ and the comments need fixing, or it cannot and should not be.

swan-amazon replied:
The comment was originally added 2 years ago. When I had left the enum as it was, I was seeing a cirque test failure because some of the commissioning tests were attempting to validate that the stage did not exceed the cleanup. After adding kConfigureTCAcknowledgments after kCleanup I was seeing this error. After I moved it before kCleanup I found that the check worked appropriately.

After making the change, we can see that the cirque tests are able to succeed after re-ordering the enum. Basically, there was some checks that depended on mutual knowledge of the (internal) enum order, between some python code and the implementati0on, but that is no longer the case.

bzbarsky-apple commented on src/controller/CommissioningDelegate.h:
```diff
@@ -581,6 +609,7 @@ class CommissioningParameters
         mAttestationNonce.ClearValue();
         mWiFiCreds.ClearValue();
         mCountryCode.ClearValue();
+        mTermsAndConditionsAcknowledgement.ClearValue();
```
Why?  This is not an "ExternalBufferDependentValue", is it?

swan-amazon replied:
You are correct. I removed the line. Now the mTermsAndConditionsAcknowledgement starts as empty Optional, and the `SetTermsAndConditionsAcknowledgement` can populate it, and re-populate it, but there is no way to return it back to Optional.

bzbarsky-apple reviewed (changes requested):
I did not review the Python bits; someone familiar with those needs to do that.

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into feature/enhanced-setup-flow-client

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -706,6 +706,25 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+     * @return CHIP_ERROR   The return status. Returns CHIP_ERROR_INCORRECT_STATE if the function is called when the device
+     *                      is not in the correct state to accept terms and conditions acknowledgements.
+     */
+    CHIP_ERROR TermsAndConditionsAcknowledgementsReady();
```
One other very important thing: this API surface would ensure that we can never do multiple commissionings in parallel, because nothing ties this call to a specific instance of the commissioning process.  This is even an issue or "serial" commissionings: one might get canceled and another started, then this call comes in for the first one....

This needs a better solution.

swan-amazon replied:
I was originally under the impression that we supported 11-digit code that did not provide VID/PID then identified the VID/PID from an announcement from the device, and then could perform a DCL lookup. I learned that we only support 21-digit and QR code for custom commissioning flow (from Chris D.). So…. We should know the VID/PID prior to initiating the AutoCommissioner, and we can require the UX to provide the acknowledgement as a pre-condition, thus avoiding any “wait” stage.

swan-amazon replied:
After discussing offline, I have removed the wait-state.

yufengwangca commented:
This PR added 14720 bytes to the fabric-sync app

swan-amazon commented:
> This PR added 14720 bytes to the fabric-sync app 

Should I file an issue against that app? Or how clusters add flash space? I'm guessing it is misconfigured?

swan-amazon committed:
Removed the "wait for terms and conditions stage"

The wait-stage is not required. The user input availability must be a
pre-condition for starting the AutoCommissioner process. The wait stage
was previously to support requesting user input after identifying the
device VID/PID using a different channel than within the pairing
payload.

swan-amazon committed:
[doc] Improve Terms and Conditions commissioning arguments documentation

Updated documentation for T&C-related commissioning arguments to better
reflect their actual usage and purpose:

- require-tc-acknowledgements: Clarified the impact on commissioning flow
- tc-acknowledgements: Explained the bit-field usage for user acceptance
- tc-acknowledgements-version: Added context about version tracking

swan-amazon committed:
[controller] Remove T&C acknowledgements from external buffer clearing

The Terms and Conditions acknowledgements parameter was incorrectly included
in ClearExternalBufferDependentValues(). This parameter contains a fixed-size
struct with two uint16_t values and does not reference any external buffers.

The CommissioningParameters class appears to be designed for additive-only
parameter setting without explicit clear/reset functionality, so removing
this inappropriate clearing operation is the correct approach.

swan-amazon committed:
[controller] Fix CommissioningStage enum order for T&C acknowledgements

Move kConfigureTCAcknowledgments before kCleanup in the CommissioningStage
enum to fix cirque test failures. The tests validate that commissioning stages
do not exceed kCleanup, which was causing failures when T&C acknowledgements
were positioned after it.

While the original comment from 2 years ago suggested the enum order was
fixed, testing reveals that the stages can be reordered. The cirque tests
now pass with this corrected ordering, indicating that any previous
constraints on enum ordering no longer apply.

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into feature/enhanced-setup-flow-client

bzbarsky-apple commented on examples/chip-tool/commands/pairing/PairingCommand.h:
```diff
@@ -202,6 +202,19 @@ class PairingCommand : public CHIPCommand,
+
+            AddArgument("require-tc-acknowledgements", 0, 1, &mRequireTCAcknowledgements,
+                        "Indicates whether Terms and Conditions acknowledgements are required during commissioning. If set to "
+                        "true, the commissioner must provide valid T&C acknowledgements or the commissioning will fail. If false, "
```
```suggestion
                        "true, the tc-acknowledgements and tc-acknowledgements-version arguments must be provided for the commissioning to succeed. If false, "
```

swan-amazon replied:
Applied after reformatting line length.

bzbarsky-apple reviewed (approved):
Approved, again with the caveat that I did not review the Python changes.

swan-amazon committed:
[doc] Clarify required arguments for T&C acknowledgements

Update help text for require-tc-acknowledgements to explicitly state which
arguments must be provided when T&C acknowledgements are required. Instead
of the vague "valid T&C acknowledgements", now specifically mentions that
both tc-acknowledgements and tc-acknowledgements-version arguments are
mandatory for successful commissioning when this option is enabled.

tcarmelveilleux commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -967,6 +972,18 @@ def __init__(self, *args):
         # The named pipe name must be set in the derived classes
         self.app_pipe = None
 
+    async def commission_devices(self) -> bool:
```
Reviewing Python, will resolve soon

tcarmelveilleux commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -671,6 +672,10 @@ class MatterTestConfig:
     trace_to: List[str] = field(default_factory=list)
 
+    # Accepted Terms and Conditions if used
+    tc_version: int = None
+    tc_user_response: int = None
```
Follow-up OK: suggest `tc_version_to_simulate` and `tc_user_response_to_simulate`.

swan-amazon replied:
Updated

tcarmelveilleux commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -967,6 +972,18 @@ def __init__(self, *args):
+        conf = self.matter_test_config
+
+        for commission_idx, node_id in enumerate(conf.dut_node_ids):
+            logging.info("Starting commissioning for root index %d, fabric ID 0x%016X, node ID 0x%016X" %
+                         (conf.root_of_trust_index, conf.fabric_id, node_id))
+            logging.info("Commissioning method: %s" % conf.commissioning_method)
```
```suggestion
            logging.info(f"Starting commissioning for root index {conf.root_of_trust_index}, fabric ID 0x{conf.fabric_id:016X}, node ID 0x{node_id:016X}")
            logging.info(f"Commissioning method: {conf.commissioning_method}")
```

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -3207,6 +3223,39 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
+            return;
+        }
+
+        GeneralCommissioning::Commands::SetTCAcknowledgements::Type request;
+        TermsAndConditionsAcknowledgement termsAndConditionsAcknowledgement = params.GetTermsAndConditionsAcknowledgement().Value();
+        request.TCUserResponse = termsAndConditionsAcknowledgement.acceptedTermsAndConditions;
+        request.TCVersion      = termsAndConditionsAcknowledgement.acceptedTermsAndConditionsVersion;
+
+        ChipLogProgress(Controller, "Setting Terms and Conditions: %hu, %hu", request.TCUserResponse, request.TCVersion);
+        CHIP_ERROR err =
+            SendCommissioningCommand(proxy, request, OnSetTCAcknowledgementsResponse, OnBasicFailure, endpoint, timeout);
```
Follow-up: consider having a separation that allows commissioning API callers to seek the correct data to use based on their external access to DCL, and continue commissioning when the terms provided by users have been received. Current impl requires this to be known a-priori, but the device being commissioned may not be known a-priori since many flows (e.g. manual pairing code) start without knowing the device type involved.

swan-amazon replied:
I had this in the previous revision, but... Boris asked to remove it, due to unnecessary complexity. Also... there is more to the story. Here is the comment I provided above.

> I was originally under the impression that we supported 11-digit code that did not provide VID/PID then identified the VID/PID from an announcement from the device, and then could perform a DCL lookup. I learned that we only support 21-digit and QR code for custom commissioning flow (from Chris D.). So…. We should know the VID/PID prior to initiating the AutoCommissioner, and we can require the UX to provide the acknowledgement as a pre-condition, thus avoiding any “wait” stage.

swan-amazon committed:
Update src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py

Co-authored-by: Tennessee Carmel-Veilleux <tennessee.carmelveilleux@gmail.com>

swan-amazon committed:
Merge branch 'master' into feature/enhanced-setup-flow-client

swan-amazon committed:
Restyle

swan-amazon committed:
Renamed variable

ksperling-apple commented on src/controller/CommissioningDelegate.h:
```diff
@@ -340,6 +353,19 @@ class CommissioningParameters
         return *this;
     }
 
+    CommissioningParameters & SetRequireTermsAndConditionsAcknowledgement(bool requireTermsAndConditionsAcknowledgement)
```
This API is confusing. Isn't it the device that *requires* (or not) TC acknowledgement or not, and this parameter controls whether or not the commissioner *should perform* TC acknowledgement?

swan-amazon replied:
I don't think the language is necessarily inappropriate. From the prospective of calling a commissioner to require that terms and conditions are sent on behalf of the client. Do you want to suggest an alternate name? Do you prefer language like "SetTermsAndConditionsUserRepsonse"?

swan-amazon replied:
I created an issue to track the api removal request. https://github.com/project-chip/connectedhomeip/issues/36965
